core/src/test/java/org/jdeferred/impl/ExceptionsTest.java

```diff
@@ -0,0 +1,145 @@
+
+        }
+    }
+
+    @Test
+    public void testFailureExceptionsAreCaught() throws InterruptedException {
+        final DeferredObject<String, Void, Void> deferredObject =
+            new DeferredObject<String, Void, Void>();
+        Promise<String, Void, Void> promise = deferredObject.promise()
+            .except(new ExceptCallback() {
+                @Override
+                public void onException(Exception e) {
+                    assertEquals(THIS_IS_MY_RUNTIME_EXCEPTION, e.getMessage());
+                }
+            })
+            .fail(new FailCallback<Void>() {
+                @Override
+                public void onFail(Void result) {
+                    throw new RuntimeException(THIS_IS_MY_RUNTIME_EXCEPTION);
+                }
+            });
+        deferredObject.reject(null);
+
+        try {
+            promise.waitSafely();
+            fail();
+        } catch (RuntimeException re) {
+        }
+    }
+
+    @Test
+    public void testAlwaysExceptionsAreCaught() throws InterruptedException {
+        final DeferredObject<String, Void, Void> deferredObject =
+            new DeferredObject<String, Void, Void>();
+        Promise<String, Void, Void> promise = deferredObject.promise()
+            .except(new ExceptCallback() {
+                @Override
+                public void onException(Exception e) {
+                    assertEquals(THIS_IS_MY_RUNTIME_EXCEPTION, e.getMessage());
+                }
+            })
+            .always(new AlwaysCallback<String, Void>() {
+                @Override
+                public void onAlways(
+                    Promise.State state, String resolved, Void rejected
+                ) {
+                    throw new RuntimeException(THIS_IS_MY_RUNTIME_EXCEPTION);
+                }
+            });
+
+        deferredObject.reject(null);
+
+        try {
+            promise.waitSafely();
+            fail();
+        } catch (RuntimeException re) {
+        }
+    }
+
+    @Test
+    public void testProgressExceptionsAreCaught() throws InterruptedException {
+        final DeferredObject<String, Void, Void> deferredObject =
+            new DeferredObject<String, Void, Void>();
+        Promise<String, Void, Void> promise = deferredObject.promise()
+            .except(new ExceptCallback() {
+                @Override
+                public void onException(Exception e) {
+                    assertEquals(THIS_IS_MY_RUNTIME_EXCEPTION, e.getMessage());
+                }
+            })
+            .progress(new ProgressCallback<Void>() {
+                @Override
+                public void onProgress(Void progress) {
+                    throw new RuntimeException(THIS_IS_MY_RUNTIME_EXCEPTION);
+                }
+            });
+        deferredObject.notify(null);
+        deferredObject.reject(null);
+
+        try {
+            promise.waitSafely();
+            fail();
+        } catch (RuntimeException re) {
+        }
+    }
+
```

Comment:
OK, I think I've got it now. The idea of this feature is to handle *runtime* exceptions that may occur inside `done`/`fail`/`always`/`progress` callbacks. All of these callbacks are executed in a background thread, provided by the `ExecutorService` used by `DeferredManager`.

I'll prefer the usage of an `Thread.UncaughtExceptionHandler` instead, however that's not as straight forward as it seems because the running thread will swallow the exception due to the inner workings of `ExecutorService`.
 There are two problems to be solved if so:
1. All places where an exception is caught and forgotten must be updated to throw a `RuntimeException`.
2. `DeferredManager` must be sure to handle exceptions to an *exception handler*, such as shown at https://github.com/griffon/griffon/blob/development/subprojects/griffon-core/src/main/java/org/codehaus/griffon/runtime/core/event/AbstractEventRouter.java#L101-L112

The *exception handler* is the `ExceptCallback` in this case.